File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.java
import java.util.HashSet;
import java.util.Set;

class Solution {
    public int lengthOfLongestSubstring(String s) {
        Set<Character> hashSet = new HashSet<>();
        int left = 0, maxLength = 0;
        for (int right = 0; right < s.length(); right++) {
            char current = s.charAt(right);
            while (hashSet.contains(current)) {
                hashSet.remove(s.charAt(left));
                left++;
            }

            hashSet.add(current);
            maxLength = Math.max(maxLength, right - left + 1);
        }

        return maxLength;
    }
}

File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.py
class Solution(object):
    def lengthOfLongestSubstring(self, s):
        l = 0
        longest = 0
        hashSet = set()
        n = len(s)

        for r in range(n):
            while s[r] in hashSet:
                hashSet.remove(s[l])
                l += 1

            w = (r - l) + 1
            longest = max(longest, w)
            hashSet.add(s[r])
        
        return longest